src/v/ssx/async_algorithm.h

```diff
@@ -106,6 +106,10 @@ template<std::random_access_iterator I, typename Fn>
 iter_size<I> for_each_limit(const I begin, const I end, ssize_t limit, Fn f) {
     auto chunk_size = std::min(limit, end - begin);
     I chunk_end = begin + chunk_size;
+    // std::for_each discards [[nodiscard]] callable result, check explicitly
+    if (false) {
+        f(*begin);
+    }
     std::for_each(begin, chunk_end, std::move(f));
```

Comment:
i'm slightly confused about what is going on. reading the commit message, it seems like it would be sufficient to disallow this function from being used for an `f` that returns a future, which would seem like the most common mistake?

then the other thing you could do is

```
f(*begin);
std::for_each(++begin, chunk_end, std::move(f));
```

?

Reply:
The original problem is indeed mistakenly supplying a future-returning function to `ssx::async_for_each` or one of its flavours. But I see no reason to limit the fix to future-returning functions specifically, and rather do it for any functions returning with `[[nodiscard]]`. It fixes the original problem and any other potential mistakes where the function result is silently lost but is not meant to.

Moreover, I don't think it would be unreasonable to create an improved version of `std::for_each` (with this static check) in our base library and use it everywhere instead.

As for your second proposal, is it because you are worried the call is optimized out before the warning is emitted? I would imagine the warning is emitted by the frontend, while any optimization, even that simple, is in the later stages. Your suggestion will also UB on an empty chunk (I'm not sure if it's called at all in this case but I'd rather not break it).

Reply:
>As for your second proposal, is it because you are worried the call is optimized out before the warning is emitted? I would imagine the warning is emitted by the frontend, while any optimization, even that simple, is in the later stages. Your suggestion will also UB on an empty chunk (I'm not sure if it's called at all in this case but I'd rather not break it).

nah, just that `if (false)` looks weird. fine i guess if it works, just wondering if there were a better option.